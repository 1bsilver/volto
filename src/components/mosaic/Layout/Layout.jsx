/**
 * Layout container.
 * @module containers/Layout
 */

/* eslint react/prefer-stateless-function: 0 */

import React, { Component, PropTypes } from 'react';
import { EditorState, ContentState } from 'draft-js';

import { Grid } from '../../../components';

/**
 * Layout component class.
 * @class Layout
 * @extends Component
 */
export default class Layout extends Component {

  /**
   * Property types.
   * @property {Object} propTypes Property types.
   * @static
   */
  static propTypes = {
    layout: PropTypes.arrayOf(
      PropTypes.arrayOf(
        PropTypes.shape({
          width: PropTypes.number,
          content: PropTypes.string,
        }),
      ),
    ).isRequired,
  }

  /**
   * Construcor
   * @param {Object} props Properties.
   * @constructs
   */
  constructor(props) {
    super(props);
    this.state = {
      layout: this.props.layout.map((row, rowIndex) =>
        row.map((tile, tileIndex) =>
          ({
            width: tile.width,
            content: EditorState.createWithContent(ContentState.createFromText(tile.content)),
            selected: rowIndex === 0 && tileIndex === 0,
          }),
        ),
      ),
      selected: {
        row: 0,
        column: 0,
      },
    };
    this.selectTile = this.selectTile.bind(this);
    this.setTileContent = this.setTileContent.bind(this);
  }

  /**
   * Set tile content.
   * @function setTileContent
   * @param {number} row Row index.
   * @param {number} column Column index.
   * @param {Object} content New content.
   * @returns {undefined}
   */
  setTileContent(row, column, content) {
    this.setState({
      layout: this.state.layout.map((rowList, rowIndex) =>
        rowList.map((tile, tileIndex) => {
          if (rowIndex === row && tileIndex === column) {
            return {
              ...tile,
              content,
            };
          }
          return tile;
        }),
      ),
    });
  }

  /**
   * Select tile.
   * @function selectTile
   * @param {number} row Row index.
   * @param {number} column Column index.
   * @returns {undefined}
   */
  selectTile(row, column) {
    this.state.layout[this.state.selected.row][this.state.selected.column].selected = false;
    this.state.layout[row][column].selected = true;
    this.setState({
      selected: {
        row,
        column,
      },
      layout: this.state.layout,
    });
  }

  /**
   * Render method.
   * @function render
   * @returns {string} Markup of the container.
   */
  render() {
    const { layout } = this.state;
    const rows = layout.map(row =>
      row.map(tile =>
        ({
          width: tile.width,
          content: tile.content,
          selected: tile.selected,
        }),
      ),
    );

    return (
      <Grid
        rows={rows}
        selectTile={this.selectTile}
        setTileContent={this.setTileContent}
      />
    );
  }
}
