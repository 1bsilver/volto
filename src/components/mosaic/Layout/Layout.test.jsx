import React from 'react';
import renderer from 'react-test-renderer';
import Layout from './Layout';

jest.mock('../Grid/Grid', () => jest.fn(() => <div />));

global.__SERVER__ = true; // eslint-disable-line no-underscore-dangle

test('renders a layout component', () => {
  const component = renderer.create(
    <Layout
      layout={[
        [
          {
            width: 12,
            content: '<h1>Hello World!',
            url: 'http://item',
          },
        ],
      ]}
    />,
  );
  const json = component.toJSON();
  expect(json).toMatchSnapshot();
});
