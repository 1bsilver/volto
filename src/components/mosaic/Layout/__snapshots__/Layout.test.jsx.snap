// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a layout component 1`] = `<div />`;
